### PR TITLE
Improve smart config generation

### PR DESCRIPTION
**Pull request by ErikWitt**
* Add cdn domains
* Only add domains and urls to whitelist if they are not covered by host regex
* Add domains where only the top level domains differs from the tested host
* Blacklist all urls that have `callback=` in their query parameters
* Restrict size of config to prevent errors on webpagetest API

### CODE REVIEW

File: baqend/_ConfigBuilder.ts [hunk 1]
```diff
@@ -26,6 +33,21 @@ export class ConfigBuilder {
     return config
   }
 
+  /**
+   * Returns the size of the config counting all included conditions.
+   */
+  getSize(): number {
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: I'd prefer using
```typescript
private _size: number
/* ... */
get size(): number { /* ... */ }
```
so …

File: baqend/_ConfigBuilder.ts [hunk 2]
```diff
@@ -26,6 +33,21 @@ export class ConfigBuilder {
...
+    return this.size
+  }
+
+  capacityReached(): boolean {
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: … you could use a getter here as well:

```js
get isCapacityReached(): boolean { /* ... */ }
```

File: baqend/_ConfigBuilder.ts [hunk 3]
```diff
@@ -43,6 +65,11 @@ export class ConfigBuilder {
   }
 
   private addToWhitelist(section: 'host' | 'pathname' | 'url', value: Condition): this {
+    if (this.capacityReached()) {
+      return this
+    }
+    this.size++
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: please use `this.size += 1` for clarity

File: baqend/_ConfigBuilder.ts [hunk 4]
```diff
@@ -71,6 +98,11 @@ export class ConfigBuilder {
   }
 
   private addToBlacklist(section: 'host' | 'pathname' | 'url', value: Condition): this {
+    if (this.capacityReached()) {
+      return this
+    }
+    this.size++
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: see above

File: baqend/_ConfigGenerator.ts [hunk 5]
```diff
@@ -168,10 +188,20 @@ export class ConfigGenerator {
...
+    return /([\w-]+\.)[\w-]+$/.test(domain) ? new RegExp(`${escapeRegExp(RegExp.$1)}[\\w-]+$`)
+      : dollarRegExp(toRegExp(domain))
+  }
+
   /**
    * Determines whether a resource is a jQuery callback.
    */
   private isJQueryCallback(resource: PuppeteerResource): boolean {
-    return !!resource.url.match(/\?callback=jQuery/)
+    return !!resource.url.match(/[\?,\&]callback=/)
```
Comment (ksm2) [inline baqend/_ConfigGenerator.ts]: Is the comma in the regex intended?

File: baqend/_ConfigGenerator.ts [hunk 6]
```diff
@@ -168,10 +188,20 @@ export class ConfigGenerator {
...
   }
 
+  /**
+   * Returns a regex that matches all subdomains of the given domain as well as other top level domains.
+   * @param {string} domain The domain to match.
+   * @return {RegExp} The regex that matches all subdomains of the given domain as well as other top level domains.
+   */
+  private matchOtherTopLevelDomains(domain: string): RegExp {
+    return /([\w-]+\.)[\w-]+$/.test(domain) ? new RegExp(`${escapeRegExp(RegExp.$1)}[\\w-]+$`)
+      : dollarRegExp(toRegExp(domain))
```
Comment (ksm2) [inline baqend/_ConfigGenerator.ts]: Please use an `if` here

File: baqend/_ConfigBuilder.ts [hunk 7]
```diff
@@ -7,9 +7,16 @@ export class ConfigBuilder {
...
   private blacklist: Rule[] = []
 
-  constructor(appName: string, userAgentDetection: boolean) {
+  /**
+   * The current size of the config counting all included conditions.
+   */
+  private size: number = 0
+  private maxCapacity: number
+
+  constructor(appName: string, userAgentDetection: boolean, maxCapacity: number = -1) {
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: You should put `CONFIG_MAX_SIZE` (see below) in __ConfigBuilder.ts_, export it, and use it as `maxCapacity`'s default:
```js
constructor(appName: string, userAgentDetection: boolean, maxCapacity: number = CONFIG_MAX_SIZE) {
```

File: baqend/_ConfigGenerator.ts [hunk 8]
```diff
@@ -9,6 +9,7 @@ import { PuppeteerResource } from './_Puppeteer'
 import credentials from './credentials'
 
 export const CDN_LOCAL_URL = 'https://makefast.app.baqend.com/v1/file/www/selfMaintainedCDNList'
+const CONFIG_MAX_SIZE = 50;
```
Comment (ksm2) [inline baqend/_ConfigGenerator.ts]: Remove here (see above)

File: baqend/_ConfigGenerator.ts [hunk 9]
```diff
@@ -47,30 +48,57 @@ export class ConfigGenerator {
...
+   * @param {boolean} mobile true if it is a mobile test, false otherwise.
+   * @param {boolean} thirdParty true if third party domains should be added to the config, false otherwise.
+   * @param {string} host The host of the tested site.
+   * @param {string[]} domains The domains loaded by the tested site.
+   * @param {PuppeteerResource[]} resources The resources loaded by the tested site.
+   * @return {Promise<Config>} A smart config for Speed Kit.
    */
   async generateSmart(url: string, mobile: boolean, thirdParty: boolean, { host, domains, resources }: { host: string, domains: string[], resources: PuppeteerResource[] }): Promise<Config> {
-    const configBuilder = new ConfigBuilder(credentials.app, mobile)
+    const configBuilder = new ConfigBuilder(credentials.app, mobile, CONFIG_MAX_SIZE)
```
Comment (ksm2) [inline baqend/_ConfigGenerator.ts]: Use as default param (revert)

File: baqend/_ConfigGenerator.ts (hunk 5 above)
Comment (ErikWitt) [reply baqend/_ConfigGenerator.ts]: No, it was not

File: baqend/_ConfigBuilder.ts [hunk 10]
```diff
@@ -36,16 +38,16 @@ export class ConfigBuilder {
...
-  getSize(): number {
-    return this.size
+  get size(): number {
+    return this._size
   }
 
-  capacityReached(): boolean {
-    if (this.maxCapacity === -1) {
+  get isCapacityReached(): boolean {
+    if (this._maxCapacity === -1) {
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: Why "-1" when there is "Infinity"?

File: baqend/_ConfigBuilder.ts [hunk 11]
```diff
@@ -65,10 +67,10 @@ export class ConfigBuilder {
   }
 
   private addToWhitelist(section: 'host' | 'pathname' | 'url', value: Condition): this {
-    if (this.capacityReached()) {
+    if (this.isCapacityReached) {
       return this
     }
-    this.size++
+    this._size = this._size + 1
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: this._size += 1

File: baqend/_ConfigBuilder.ts [hunk 12]
```diff
@@ -98,10 +100,10 @@ export class ConfigBuilder {
   }
 
   private addToBlacklist(section: 'host' | 'pathname' | 'url', value: Condition): this {
-    if (this.capacityReached()) {
+    if (this.isCapacityReached) {
       return this
     }
-    this.size++
+    this._size = this._size + 1
```
Comment (ksm2) [inline baqend/_ConfigBuilder.ts]: see above

File: baqend/_ConfigGenerator.ts [hunk 13]
```diff
@@ -194,14 +193,17 @@ export class ConfigGenerator {
    * @return {RegExp} The regex that matches all subdomains of the given domain as well as other top level domains.
    */
   private matchOtherTopLevelDomains(domain: string): RegExp {
-    return /([\w-]+\.)[\w-]+$/.test(domain) ? new RegExp(`${escapeRegExp(RegExp.$1)}[\\w-]+$`)
-      : dollarRegExp(toRegExp(domain))
+    if (/([\w-]+\.)[\w-]+$/.test(domain)) {
+      return  new RegExp(`${escapeRegExp(RegExp.$1)}[\\w-]+$`)
```
Comment (ksm2) [inline baqend/_ConfigGenerator.ts]: double space here

**Pull request merged**